src/submodule.c

```diff
@@ -688,6 +691,34 @@ int git_submodule_add_setup(
 		goto cleanup;
 	}
 
+	/* get the index for the repo */
+
+	if ((error = git_repository_index__weakptr(&index, repo)) < 0)
+		goto cleanup;
+
+	/* see if the submodule name exists as a file on the index */
+
+	if ((error = git_index_find(NULL, index, path)) == 0) {
+		giterr_set(GITERR_SUBMODULE,
+			"'%s' already exists in the index", path);
+		return GIT_EEXISTS;
+	}
+
+	/* We need the path to end with '/' so we can check it as a directory prefix */
+
+	path_len = strlen(path);
+	dir = git__malloc(path_len + 1);
+	strcpy(dir, path);
+	git_path_string_to_dir(dir, path_len + 1);
```

Comment:
You could use a buffer here with `git_path_to_dir`:

```
git_buf path = GIT_BUF_INIT;
git_buf_sets(&path, path);
git_path_to_dir(&path);
```

Error handling is obviously missing.

Reply:
The code is obviously bogus, I should've chosen two different variable names. But I guess you get what I mean :P